File: packages/secret-guard/dart/test/sinks_test.dart
import 'package:flutter_test/flutter_test.dart';
import 'package:secret_guard/secret_guard.dart';

void main() {
  setUp(() {
    resetRegistryForTests();
  });

  test('registerSink returns a mask()-backed function usable at the call site',
      () {
    final maskIt =
        registerSink(const SinkRegistration(SinkKind.log, 'test-call-site'));
    final out = maskIt('token=${'a' * 32}');
    expect(out.contains('a' * 32), isFalse);
  });

  test('records the registration for later coverage checks', () {
    registerSink(const SinkRegistration(SinkKind.log, 'test-call-site'));
    final sinks = listRegisteredSinks();
    expect(sinks.length, 1);
    expect(sinks[0].kind, SinkKind.log);
    expect(sinks[0].name, 'test-call-site');
  });

  test('throws on duplicate (kind, name) registration', () {
    registerSink(const SinkRegistration(SinkKind.log, 'dup'));
    expect(
      () => registerSink(const SinkRegistration(SinkKind.log, 'dup')),
      throwsA(isA<DuplicateSinkRegistrationError>()),
    );
  });

  test('allows the same name under a different kind', () {
    registerSink(const SinkRegistration(SinkKind.log, 'shared-name'));
    expect(
      () => registerSink(
        const SinkRegistration(SinkKind.httpResponse, 'shared-name'),
      ),
      returnsNormally,
    );
  });

  group('assertAllKindsRegistered — the 未登録経路検出 test', () {
    test('throws listing every missing kind when nothing is registered', () {
      expect(
        () => assertAllKindsRegistered(),
        throwsA(isA<MissingSinkKindsError>().having(
          (e) => e.missing.length,
          'missing.length',
          allSinkKinds.length,
        )),
      );
    });

    test('throws only for the kinds still missing after partial registration',
        () {
      registerSink(const SinkRegistration(SinkKind.log, 'a'));
      registerSink(const SinkRegistration(SinkKind.httpResponse, 'b'));
      try {
        assertAllKindsRegistered();
        fail('expected assertAllKindsRegistered to throw');
      } on MissingSinkKindsError catch (e) {
        expect(e.missing, contains(SinkKind.errorMessage));
        expect(e.missing, contains(SinkKind.urlQuery));
        expect(e.missing, contains(SinkKind.artifactFile));
        expect(e.missing, isNot(contains(SinkKind.log)));
        expect(e.missing, isNot(contains(SinkKind.httpResponse)));
      }
    });

    test('passes once every kind has at least one registered sink', () {
      for (final kind in allSinkKinds) {
        registerSink(SinkRegistration(kind, 'synthetic-${kind.name}'));
      }
      expect(() => assertAllKindsRegistered(), returnsNormally);
    });
  });
}

File: packages/secret-guard/src/__tests__/sinks.test.ts
import { describe, it, expect, beforeEach } from "vitest";
import {
  registerSink,
  listRegisteredSinks,
  assertAllKindsRegistered,
  ALL_SINK_KINDS,
  _resetRegistryForTests,
} from "../sinks";

describe("registerSink() / sink registry", () => {
  beforeEach(() => {
    _resetRegistryForTests();
  });

  it("returns a mask()-backed function usable at the call site", () => {
    const maskIt = registerSink({ kind: "log", name: "test-call-site" });
    const out = maskIt("token=" + "a".repeat(32));
    expect(out).not.toContain("a".repeat(32));
  });

  it("records the registration for later coverage checks", () => {
    registerSink({ kind: "log", name: "test-call-site" });
    const sinks = listRegisteredSinks();
    expect(sinks).toHaveLength(1);
    expect(sinks[0]).toMatchObject({ kind: "log", name: "test-call-site" });
  });

  it("throws on duplicate (kind, name) registration", () => {
    registerSink({ kind: "log", name: "dup" });
    expect(() => registerSink({ kind: "log", name: "dup" })).toThrow(
      /already registered/
    );
  });

  it("allows the same name under a different kind (kind is part of the identity)", () => {
    registerSink({ kind: "log", name: "shared-name" });
    expect(() =>
      registerSink({ kind: "http_response", name: "shared-name" })
    ).not.toThrow();
  });

  describe("assertAllKindsRegistered() — the 未登録経路検出 (undetected-route) test", () => {
    it("throws listing every missing kind when nothing is registered", () => {
      expect(() => assertAllKindsRegistered()).toThrow(
        new RegExp(ALL_SINK_KINDS.join(".*"))
      );
    });

    it("throws only for the kinds still missing after partial registration", () => {
      registerSink({ kind: "log", name: "a" });
      registerSink({ kind: "http_response", name: "b" });
      try {
        assertAllKindsRegistered();
        throw new Error("expected assertAllKindsRegistered to throw");
      } catch (err) {
        const message = (err as Error).message;
        expect(message).toContain("error_message");
        expect(message).toContain("url_query");
        expect(message).toContain("artifact_file");
        expect(message).not.toContain("log,");
      }
    });

    it("passes once every kind has at least one registered sink", () => {
      for (const kind of ALL_SINK_KINDS) {
        registerSink({ kind, name: `synthetic-${kind}` });
      }
      expect(() => assertAllKindsRegistered()).not.toThrow();
    });
  });
});
